sling/nlp/parser/trainer/transition-generator.cc

```diff
@@ -0,0 +1,457 @@
+    case ParserAction::EMBED:
+      output.label = action.frame->type;
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ELABORATE:
+      output.label = action.frame->type;
+      output.source = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::CONNECT:
+      output.source = Index(attention, action.frame->handle);
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ASSIGN:
+      output.source = Index(attention, action.frame->handle);
+      output.label = action.core.label;
+      break;
+    default:
+    break;
+  }
+
+  return output;
+}
+
+// Updates 'attention' as a result of execution 'action'.
+void Update(const Action &action, Handles *attention) {
+  auto type = action.core.type;
+  if (type == ParserAction::EVOKE || type == ParserAction::EMBED ||
+      type == ParserAction::ELABORATE) {
+    attention->insert(attention->begin(), action.frame->handle);
+  } else if (type == ParserAction::REFER || type == ParserAction::ASSIGN ||
+             type == ParserAction::CONNECT) {
+    std::remove(attention->begin(), attention->end(), action.frame->handle);
+    attention->insert(attention->begin(), action.frame->handle);
+  }
+}
+
+void InitInfo(
+    Frame frame, FrameInfo *frame_info, HandleSet *initialized) {
+  Handle handle = frame.handle();
+  if (initialized->find(handle) != initialized->end()) return;
+  initialized->insert(handle);
+
+  auto *entry = frame_info->Get(handle);
+  std::vector<Frame> pending;
+  for (const Slot &slot : frame) {
+    if (slot.name.IsId()) continue;
+
+    Frame value(frame.store(), slot.value);
+    if (!value.IsFrame()) continue;
+
+    if (slot.name.IsIsA() && entry->type.IsNil()) {
+      entry->type = slot.value;
+    } else {
+      auto *edge = new FrameInfo::Edge(slot.name, false, slot.value);
+      entry->edges.emplace_back(edge);
+      if (slot.value == handle) {
+        edge->inverse = edge;
+        continue;
+      }
+      if (value.IsLocal()) {
+        auto *neighbor = frame_info->Get(slot.value);
+        auto *reverse = new FrameInfo::Edge(slot.name, true, handle);
+        neighbor->edges.emplace_back(reverse);
+        reverse->inverse = edge;
+        edge->inverse = reverse;
+        pending.emplace_back(value);
+      }
+    }
+  }
+
+  for (Frame f : pending) {
+    InitInfo(f, frame_info, initialized);
+  }
+}
+
+void InitializeGeneration(
+    const Document &document,
+    int begin, int end,
+    FrameInfo *frame_info,
+    std::vector<TokenToMentions> *token_to_mentions) {
+  Handle h_evoked = document.store()->LookupExisting("evokes");
+  QCHECK(!h_evoked.IsNil());
+  HandleSet initialized;
+  token_to_mentions->resize(end - begin);
+
+  for (const Span *span : document.spans()) {
+    if (span->deleted() || span->begin() < begin || span->end() > end) continue;
+
+    for (const Slot &slot : span->mention().Slots(h_evoked)) {
```

Comment:
It might be simpler to use span->AllEvoked() here.

Reply:
Done